src/perform_step/multirate_perform_step.jl

```diff
@@ -0,0 +1,175 @@
+get_a21(tab::OrdinaryDiffEqTableau{T,T2}) where {T,T2} = hasfield(typeof(tab), :a21) ? tab.a21 : zero(T)
```

Comment:
hmm, interesting. This won't scale... but does it need to?

Reply:
we probably wont need more than 4-5 order methods, base methods are usually small order ones

Reply:
why only low order? Why does this not apply to arbitrary methods?

Reply:
Maybe one could define something like
```julia
struct Zero end

function Base.getproperty(tab::OrdinaryDiffEqTableau, sym::Symbol)
    if hasfield(tab, sym)
        getfield(tab, sym)
    else
        Zero()
    end
end

Base.:+(x::Real, ::Zero) = x
Base.:+(::Zero, x::Real) = x
Base.:+(::Zero, ::Zero) = Zero()

Base.:*(x::Real, ::Zero) = Zero()
Base.:*(x::Zero, y::Real) = y * x
Base.:*(::Zero, ::Zero) = Zero()
```
and the compiler would be smart enough to compile away all the unneeded operations if the fields don't exist. 

The annoying thing about just returning `zero(T)` is that all the multiplications and additions will still be performed. Additionally, in this way you could still just use `@unpack a21, .... = cache.tab` instead of defining all these (arguably) ugly `get_...` methods.

Reply:
> why only low order? Why does this not apply to arbitrary methods?  
Not sure but Sandu's paper suggests the same that base methods are generally small order ones

> and the compiler would be smart enough to compile away all the unneeded operations if the fields don't exist.   
This is nice idea, would this be compatible with muladd too?

Reply:
> This is nice idea, would this be compatible with muladd too?

Probably not with the code snippet above, but you could just define dispatches for `Base.muladd`.

BTW I actually wanted to write "and hopefully the compiler would be smart enough" since I didn't check it carefully, I just assumed it would. I did that now, and it seems to work in the following example:
```julia
struct Tableau{A}
    a::A
end

struct Tableau2{A,B}
    a::A
    b::B
end

function f(t)
    a = t.a
    b = t.b
    return a + 42 * b * (500 + 3.0)
end

@code_typed f(Tableau(1.0))
@code_llvm f(Tableau(1.0))
@code_typed f(Tableau2(1.0, 2.0))
@code_llvm f(Tableau2(1.0, 2.0))
```

Reply:
The following doesn't compile away though (unsurprisingly, I guess):
```julia
function f(t)
    a = t.a
    b = t.b
    return a + 42 * b * rand()
end
```
For broadcasting support one has to add some additional definitions, it seems.

Reply:
BTW `muladd` seems to work for free due to the fallback in https://github.com/JuliaLang/julia/blob/30b09dfd2b3a898237f71d4a2b0812466d017f9e/base/math.jl#L1139.

Reply:
>BTW muladd seems to work for free due to the fallback in https://github.com/JuliaLang/julia/blob/30b09dfd2b3a898237f71d4a2b0812466d017f9e/base/math.jl#L1139.

But that means it's not actually doing muladd, which isn't what we want.

Reply:
But isn't that what we want here? We don't want any computations at all since we just add zero or multiply with zero. So the fallback is fine since it allows use to eliminate them at compile time.

Reply:
I guess so.

Reply:
I think this approach is good now. But we need to make sure we implement it right, like return `Zero()` only if the requested field is of form `aXY`, `bX`, `cX` etc otherwise debugging stuff won't be easy lol

Reply:
Good for next PR or do it in the same?

Reply:
yeah